crates/futures/src/queue.rs

```diff
@@ -0,0 +1,90 @@
+use js_sys::Promise;
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+use std::rc::Rc;
+use wasm_bindgen::prelude::*;
+
+pub(crate) trait Task {
+    fn run(self);
+}
+
+struct QueueState<A> {
+    // This is used to ensure that it's only scheduled once
+    is_spinning: Cell<bool>,
+
+    // This is a queue of Tasks which will be run in order
+    tasks: RefCell<VecDeque<A>>,
+}
+
+impl<A> QueueState<A>
+where
+    A: Task,
+{
+    fn run_all(&self) {
+        loop {
+            // We immediately drop the borrow_mut because `run` might queue more tasks
+            let task = self.tasks.borrow_mut().pop_front();
+
+            match task {
+                Some(task) => {
+                    task.run();
+                }
+                None => {
+                    // All of the Tasks have been run, so it's now possible to schedule the next tick again
+                    self.is_spinning.set(false);
+                    break;
+                }
+            }
+        }
+    }
+}
+
+pub(crate) struct Queue<A> {
+    state: Rc<QueueState<A>>,
+    promise: Promise,
+    closure: Closure<dyn FnMut(JsValue)>,
+}
+
+impl<A> Queue<A> {
+    pub(crate) fn push_task(&self, task: A) {
+        let mut lock = self.state.tasks.borrow_mut();
+
+        lock.push_back(task);
+
+        // If we already scheduled the next tick then do nothing
+        if self.state.is_spinning.replace(true) {
+            return;
+        }
+
+        // The Task will be run on the next microtask event tick
+        // TODO replace with https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/queueMicrotask
```

Comment:
I think it's fine to remove the "TODO" here and basically just comment that this is probably slightly more expensive than it needs to be and the proposed API here may be usable instead. I doubt we'll ever need to do it though since this is already so heavily amortized.

Reply:
I definitely plan on replacing it with `queueMicrotask` once it stabilizes. I use TODOs to basically mean, "there is something we can't do now, but we should do it later", just as a reminder so I (or others) don't forget.